### PR TITLE
dwarf: dedup generation of dwarf info for func args and vars in Dwarf module

### PR DESCRIPTION
**Pull request by kubkon**
I think it's high time we move all DWARF specific stuff such as opcodes and exprloc emitting to `Dwarf.zig` module and expose higher level helpers for use by any codegen that wants to emit DWARF debug info.

cc @joachimschmidt557, @Luukdegram, @koachan

### CODE REVIEW

**Comment by Luukdegram** [comment]
While I'm a fan of the deduplication of the logic, I wonder if this may also add duplication in other contexts now. For instance, each time we add an extra variant to `VarArgDbgInfoLoc` we have to update all backends that are switching on it. For instance, I had only implemented locals for the WebAssembly target so far, but there are going to be more 'custom' ones. Perhaps having the logic of the main opcodes deduplicated, as this PR does, far outweighs the concern I just mentioned though. Happy to hear what you think about this. I do think this makes it much, much easier for conventional targets to support DWARF debug info without too much work. It also means debug logic is contained within the `Dwarf` module, and not in the backend as that should perhaps only be concerned about generating code, rather than debug information.

**Comment by kubkon** [comment]
> While I'm a fan of the deduplication of the logic, I wonder if this may also add duplication in other contexts now. For instance, each time we add an extra variant to `VarArgDbgInfoLoc` we have to update all backends that are switching on it. For instance, I had only implemented locals for the WebAssembly target so far, but there are going to be more 'custom' ones. Perhaps having the logic of the main opcodes deduplicated, as this PR does, far outweighs the concern I just mentioned though. Happy to hear what you think about this. I do think this makes it much, much easier for conventional targets to support DWARF debug info without too much work. It also means debug logic is contained within the `Dwarf` module, and not in the backend as that should perhaps only be concerned about generating code, rather than debug information.

The backends don't switch on `VarDbgInfoLoc` though - they only ever generate it. That's the whole point of this abstraction, converting backend-specific `MCValue`s into a common set of DWARF location encodings.

**Comment by kubkon** [comment]
I should probably also add that now, if you add a new Wasm-specific mapping between a WasmValue and a DWARF location, all you need to do is add a new variant to `genArgDbgInfo` and `genVarDbgInfo`  and implement them (in `Dwarf.DeclState`). Other backends don't care what the new variant is and so the change is completely transparent to them. That is my main goal of this refactoring at least and this is how it should work. Does it sound more convincing now to you @Luukdegram? :-)

**Comment by Luukdegram** [comment]
Sounds excellent!

File: src/link/Dwarf.zig [hunk 1]
```diff
@@ -560,6 +561,222 @@ pub const DeclState = struct {
...
     }
+
+    pub fn genArgDbgInfo(
+        self: *DeclState,
+        name: [:0]const u8,
+        ty: Type,
+        atom: *Atom,
+        loc: union(enum) {
+            register: u8,
+            stack: struct { fp_register: u8, offset: i32 },
```
Comment (joachimschmidt557) [inline src/link/Dwarf.zig:572]: For the future we should also consider using the stack pointer as the base register (in order to relieve register pressure when it's ok to omit the frame pointer)

**Review by joachimschmidt557 — APPROVED** [review]
Looks great!

File: src/link/Dwarf.zig (hunk 1 above)
Comment (kubkon) [reply src/link/Dwarf.zig:572]: Yep, totally agree. There is also a CLI flag to that effect we will need to respect `-fno-frame-pointer`.

Comment (joachimschmidt557) [reply src/link/Dwarf.zig]: I'm only aware of `-fomit-frame-pointer`. As far as I understand, this only tells the compiler that it is _possible_ to omit the frame pointer, not a rigid requirement.

```
joachim@dustbowl ~> zig build-exe --help | grep frame-pointer
  -fomit-frame-pointer      Omit the stack frame pointer
  -fno-omit-frame-pointer   Store the stack frame pointer
```

Comment (kubkon) [reply src/link/Dwarf.zig]: Oops, that's what I meant yes.

**Pull request merged**